### ISSUE TITLE
Support for gl_PrimitiveID / SV_PrimitiveID -like construct?

### ISSUE THREAD

**Issue by darzu** [gpuweb/gpuweb#1786]
I would like to index into a uniform buffer per-primitive (e.g. triangle face) in the fragment shader, so I can have per-face data (e.g. color). I can't find any way to do this within the current WGSL spec.

**Comment by Kangz** [comment]
Unfortunately `PrimitiveID` isn't support in the WebGPU min-spec: from my reading of the Vulkan spec and MSL guide, both require a geometry or tessellation stage in the pipeline to make it work and base WebGPU can't support that.

What you could do is use a flat vertex out variable that writes a primitive ID, if you are doing a regular draw then you can just divide the vertex ID by 3, but if you're doing an indexed draw, it might require an additional vertex buffer with IDs (which is kind of unfortunate).

**Comment by kvark** [comment]
> if you're doing an indexed draw, it might require an additional vertex buffer with IDs (which is kind of unfortunate).

Worth noting that you can't "just" add this another buffer. You'd actually need to change the indices now, since neighboring polygons wouldn't be able to share a vertex. This would be equivalent to just removing the index buffer entirely, and will come with a noticeable performance hit.

Primitive ID is not available in vertex shaders because it's too expensive, basically.

**Comment by Kangz** [comment]
> Worth noting that you can't "just" add this another buffer. You'd actually need to change the indices now, since neighboring polygons wouldn't be able to share a vertex. This would be equivalent to just removing the index buffer entirely, and will come with a noticeable performance hit.

Whoops, that's right!

**Comment by darzu** [comment]
Thanks for the info! I'll likely add the extra info via the vertex buffer. Reshuffling the indices isn't too bad in my scenario, except that there are some polygons that have more faces than vertices so I'm not sure what I'll do there yet.

**Comment by martinken** [comment]
PrimitiveID also has sort of odd behavior with degenerate primitives IIRC. A triangle with three unique vertex indices (but degenerate as their positions are equal) doesn't increment PrimitiveID causing your cell array to be out of sync unless you check for degenerate primitives when building your cell array (if you have to handle degenerate primitives). 

We have the same use case. For us right now we stick the cell data/id into the vertex array and duplicate any shared vertices. In the future we plan to switch to using the flat attribute to reduce the number of duplicates. (from 6x for a closed tri mesh to ~2x)

**Comment by darzu** [comment]
Exactly my thought, I'm waiting for the flat attribute as well which as best as I can tell is being tracked here for Chrome: https://bugs.chromium.org/p/tint/issues/detail?id=746&q=interpolate&can=2

**Comment by Kangz** [comment]
Reopening as some developers were interested in such a builtin: in Vulkan turns out that `PrimitiveID` is supported in fragment shaders if the `Geometry` or `Tessellation` SPIR-V capabilities are present, themselves controlled by the respective Vulkan optional feature. So while not min-spec, it should be fairly common. 

In MSL things are less clear but it seems it may be supported as soon as MSL 2.2? (it is called `primitive_id` in MSL)

**Comment by litherum** [comment]
Does the OP want this in vertex shaders or fragment shaders?

**Comment by litherum** [comment]
Metal has `[[primitive_id]]` as input to fragment shaders, since Metal 2.3. (See page 100 https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-04-11-minutes#support-for-gl_primitiveid--sv_primitiveid--like-construct-1786">WGSL 2023-04-11 Minutes</a></summary>

* MM: At least from our perspective, a natural win. If there’s support and people want it, let’s open it up. But only in the fragment shader. In the vertex shader there’s no such thing as a primitive
* KG: Need a PR to precipitate some investigation. Seems like an easy win, but, it didn’t ship with the V1. What’s the compatibility story?
* DN: This hasn’t been on our radar, so we don’t have an opinion right now (bad audio) need more time
* AB: Let’s defer
* **DEFERRED**

</details>

**Comment by dneto0** [comment]
Alan and I have investigated:

* In Vulkan, the PrimitiveID feature is gated by geometry shaders, mesh shaders, or raytracing.  So any feature would be limited in its reach. 
* In MSL, in fragment shaders it *looks* like primitive_id is only used together with barycentric coordinates.  That doesn't seem to be as general a feature as is being requested.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-04-18-minutes#support-for-gl_primitiveid--sv_primitiveid--like-construct-1786">WGSL 2023-04-18 Minutes</a></summary>

* KG: Milestone 2.  Needs concrete PR.
* AB: We think it needs an enable.  It’s not a software feature.
* DN: In MSL it seemed tied to barycentric coordinates. Which is maybe narrower than what the original poster asked for.  Suggest needs investigation.
* KG: I think I’ll place this in “post-v1”, and someone can investigate/make a pr.
* 

</details>

**Comment by kohakukun** [comment]
We have a use case where we use the gl_primitiveID to access a buffer in the fragment shader, similar to the original use case.

**Comment by mwyrzykowski** [comment]
We had another request for this, `primitive_id` in MSL is not limited to barycentric coordinates however it is only supported by Apple7 and up (and all Macs). WebGPU core spec is currently Apple4 or 5 I think so if we wanted to add it, it could be an extension.

It appears MSL (limited to Mac or iOS Apple7+) and D3D support this (SV_PrimitiveID), so I think it meets the requirements for being an extension

**Comment by JMS55** [comment]
Not really necessary/good, leading vertex is (currently) better assuming you can pre-process your mesh data.

See slide ~72 https://www.advances.realtimerendering.com/s2024/index.html#hable

**Comment by coffeenotfound** [comment]
The leading vertex / index rotation prim id technique is quite clever but keep in mind that most normal meshes have about twice the number of triangles compared to vertices. For this trick you need at least as many vertices as triangles to store the prim ids, so you'll end up with about double the verts

**Comment by mwyrzykowski** [comment]
Recommend adding this as an API + WGSL extension

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2025-07-16#support-for-gl_primitiveid--sv_primitiveid--like-construct-1786">GPU Web WG meeting 2025-07-16</a></summary>

* CW: This is present in all APIs but sometimes gated on versions / extensions. Useful to get per-triangle data (normals) or when rendering the triangle ID for Nanite-like usecases.
* JB: Surprised that this didn't exist already because it seems so useful to reduce memory consumption.
* CW: Agreement for "PR welcome"
* JB: Optional feature to expose where it's supported?
* CW: Yes
* KN: What is needed in Vulkan for this?
* CW: Geometry, tesselation or RT, most devices.

</details>

**Comment by dj2** [comment]
We've started prototyping and testing out Primitive ID in Dawn/Tint. You can see the initial design document at [ChromiumExperimentalPrimitiveId](https://dawn.googlesource.com/dawn/+/refs/heads/main/docs/tint/extensions/chromium_experimental_primitive_id.md). Our plan is to start creating some dawn end to end tests and CTS tests around what we've got now so we can see if things work as we hope/think they should.

There are a few outstanding questions in the design which we need to resolve, but hopefully they'll shake out of the testing. Once we've confirmed the various backends work as spec'd we'll push up spec PRs for WGSL and WebGPU.

**Comment by mwyrzykowski** [comment]
Fyi as touched on in the meeting today, tessellation and mesh shading are only supported by Metal on devices where barycentric coordinates are supported.

So without complex polyfills, this needs to remain an optional extension as it requires Apple7+ (A14 chips or later)

**Comment by dj2** [comment]
@toji has the spec for this is written, the PR has editor approval. The CTS tests are well under way.

There seemed to be agreement in the previous meeting to do this from the various browsers, do we need to wait for another meeting, or can we get WG approval here from the various stakeholders to push this forward and land the spec change? (Dawn has an implementation ready to go, we just need to rename the enables from `chromium_experimental_primitive_id` to `primitive_id`).
